src/common.cpp

```diff
@@ -43,32 +44,75 @@ inline namespace MIGRAPHX_INLINE_NS {
 // In this case we need to broadcast the (:,:,1:,:) axis
 // of s0 plus the 1st dimension of s1 giving
 // output_lens = (3,2,7,5)
+//
 std::vector<std::size_t> compute_broadcasted_lens(std::vector<std::size_t> s0,
                                                   std::vector<std::size_t> s1)
 {
     if(s0 == s1)
         return s0;
     if(s0.size() > s1.size())
         s0.swap(s1);
-
     std::vector<std::size_t> out_lens(s1);
     auto offset = s1.size() - s0.size();
     std::transform(
         s0.begin(), s0.end(), s1.begin() + offset, out_lens.begin() + offset, [&](auto a, auto b) {
             if(a != b and a != 1 and b != 1)
             {
-                MIGRAPHX_THROW("COMPUTE_BROADCASTLEN: shape {" + to_string_range(s0) + "} and {" +
-                               to_string_range(s1) + "} mismatch!");
+                MIGRAPHX_THROW("COMPUTE_BROADCASTLEN: shape {" + migraphx::to_string_range(s0) +
+                               "} and {" + migraphx::to_string_range(s1) + "} mismatch!");
             }
             return std::max(a, b);
         });
-
     return out_lens;
 }
 
+std::vector<shape::dynamic_dimension> compute_broadcasted_dyn_dims(shape s0, shape s1)
+{
+    assert(s0.dynamic() or s1.dynamic());
+    // change both shapes to dynamic_dimension representation
+    if(not s0.dynamic())
+        s0 = s0.to_dynamic();
+    if(not s1.dynamic())
+        s1 = s1.to_dynamic();
+
+    if(s0.ndim() > s1.ndim())
+    {
+        std::swap(s0, s1);
+    }
+    auto offset = s1.ndim() - s0.ndim();
+    std::vector<shape::dynamic_dimension> out_dims(s1.dyn_dims());
+    std::vector<shape::dynamic_dimension> one_dyn_dims{{1, 1, 0}, {1, 1, 1}};
+    std::transform(
+        s0.dyn_dims().cbegin(),
+        s0.dyn_dims().cend(),
+        s1.dyn_dims().cbegin() + offset,
+        out_dims.begin() + offset,
+        [&](auto a, auto b) {
+            if(a == b)
+            {
+                return a;
+            }
+            else if(contains(one_dyn_dims, a) or contains(one_dyn_dims, b))
+            {
+                // setting opt to 0, may need to be changed
+                return shape::dynamic_dimension{std::max(a.min, b.min), std::max(a.max, b.max), 0};
```

Comment:
The `std::max(a.min, b.min)` seems like a typo.

We should first take the dimensions from the fixed dimension if available:

```cpp
if (a.is_fixed())
    return a;
else if (b.is_fixed())
    return b;
else
    shape::dynamic_dimension{std::min(a.min, b.min), std::max(a.max, b.max), 0}
```

But I wonder if the last `else` should maybe be an error instead.

Reply:
This is the case where one of the shapes (after it is converted to a dynamic shape if it was a static shape) has a fixed dynamic dimension of 1. In that case, it should just broadcast to the other dynamic dimension. Which `std::max` should do. I don't think it should prefer a fixed dynamic_dimension.

ex:
a = {1, 4, 2}   b = {1, 1, 0}   ->  {1, 4, 0}